### PR TITLE
plat-sam: add support for Microchip sama7g54-ek board

### PR DESCRIPTION
**Pull request by TonyHan11**

### CODE REVIEW

**Review by jenswi-linaro — COMMENTED** [review]
Each commit should build as it is added. When looking at these patches I'm not sure that's the case.
"plat-sam: add functions for support for sama7g5 initialize" adds code that isn't built yet.
"plat-sam: add support for Microchip sama7g54-ek board" breaks some or all sama5d2 builds.

File: core/arch/arm/plat-sam/platform_sama7g5.c [hunk 1]
```diff
@@ -0,0 +1,960 @@
...
+{
+	matrix_write_protect_disable(matrix_base(MATRIX_SAMA7G54));
+	matrix_configure_slave();
+}
+
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, TZC_BASE_ADDRESS, CORE_MMU_PGDIR_SIZE);
+
+static void tzc400_init(void)
+{
+	struct tzc_region_config cfg = {};
```
Comment (etienne-lms) [inline core/arch/arm/plat-sam/platform_sama7g5.c:894]: s/`{}`/`{ }`/  (with a space char in between braces)

File: core/arch/arm/plat-sam/platform_sama7g5.c [hunk 2]
```diff
@@ -0,0 +1,960 @@
...
+	unsigned int tzc_idx = 0;
+	vaddr_t addr = 0;
+
+	for (tzc_idx = 0; tzc_idx <= 1; tzc_idx++) {
+		addr = TZC_BASE_ADDRESS + 0x1000 * tzc_idx;
+		if (cpu_mmu_enabled())
+			addr = (vaddr_t)phys_to_virt(addr, MEM_AREA_IO_SEC, 1);
+		tzc_init(addr);
+
+		cfg.filters = (tzc_idx == 0 ? GENMASK_32(3, 0) : BIT(0));
```
Comment (etienne-lms) [inline core/arch/arm/plat-sam/platform_sama7g5.c:904]: nitpicking: prefer
```c
		if (tzc_idx)
			cfg.filters = BIT(0);
		else
			cfg.filters = GENMASK_32(3, 0);

File: core/arch/arm/plat-sam/conf.mk [hunk 3]
```diff
@@ -45,6 +57,14 @@ $(call force,CFG_CORE_DYN_SHM,n)
 
 CFG_MMAP_REGIONS ?= 24
 
+ifeq ($(CFG_SAMA7G5),y)
+CFG_SHMEM_START  = 0x61000000
```
Comment (etienne-lms) [inline core/arch/arm/plat-sam/conf.mk:61]: use either `CFG_xxx ?= <VALUE>` or `$(call force,CFG_xxx,<VALUE>)` for `CFG_` config switches

File: core/arch/arm/plat-sam/platform_sama7g5.c (hunk 1 above)
Comment (TonyHan11) [reply core/arch/arm/plat-sam/platform_sama7g5.c]: Updated.

File: core/arch/arm/plat-sam/platform_sama7g5.c (hunk 2 above)
Comment (TonyHan11) [reply core/arch/arm/plat-sam/platform_sama7g5.c]: Updated.

File: core/arch/arm/plat-sam/conf.mk (hunk 3 above)
Comment (TonyHan11) [reply core/arch/arm/plat-sam/conf.mk]: Updated, thanks.

**Comment by TonyHan11** [comment]
Thank you @jenswi-linaro, commits reorganized with the following updates:
 - license updated.
 - `|` moved to the end of lines.
 - all the `register_phys_mem_*` moved to before the implementation of functions
 - comment with `//` updated to `/* xxx */`.
 - as mmu is not enabled when `tzc400_init()` and `tzpm_init()` are called, remove `phys_to_virt()` and add `assert(!cpu_mmu_enabled());`
 - update conf.mk and replace `at91-common` to `src-y` in core/drivers/clk/sam/sub.mk

**Comment by TonyHan11** [comment]
Fix `src-y =` with `src-y +=` in `core/drivers/clk/sam/sub.mk` when replacing `at91-common` with `src-y`.

File: core/arch/arm/plat-sam/conf.mk [hunk 4]
```diff
@@ -45,6 +58,14 @@ $(call force,CFG_CORE_DYN_SHM,n)
...
 CFG_MMAP_REGIONS ?= 24
 
+ifeq ($(CFG_SAMA7G5),y)
+CFG_SHMEM_START ?=0x61000000
+CFG_SCMI_SHMEM_START ?= 0x61400000
+CFG_DT_ADDR ?= 0x61500000
+CFG_ATMEL_SFC ?= n
+CFG_NVMEM_DIE_ID ?= n
+CFG_NVMEM_HUK ?= n
+endif
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-sam/conf.mk:68]: I was expecting `else` here to cover all the variables above for the other case, something like:
```
else
CFG_SHMEM_START = 0x21000000
CFG_SCMI_SHMEM_START  ?= 0x21400000
...
CFG_NVMEM_HUK ?= y
endif
```
to make it easier to see the difference.

Comment (TonyHan11) [reply core/arch/arm/plat-sam/conf.mk]: Calculated the addresses in Makefile and updated for the other case, thanks.

**Comment by TonyHan11** [comment]
Reabse the commits onto the latest `master` branch.

File: core/arch/arm/plat-sam/conf.mk [hunk 5]
```diff
@@ -45,16 +58,16 @@ $(call force,CFG_CORE_DYN_SHM,n)
 
 CFG_MMAP_REGIONS ?= 24
 
-CFG_SHMEM_START  ?= 0x21000000
+CFG_SHMEM_START  ?= $(shell echo $$(($(CFG_TZDRAM_START) + 0x1000000)))
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-sam/conf.mk:61]: Doesn't `CFG_SHMEM_START  ?= ($(CFG_TZDRAM_START) + 0x1000000)` work?
If it does, please use that pattern instead, here and below.

**Review by jenswi-linaro — COMMENTED** [review]
With my comment addressed, please apply:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/plat-sam/conf.mk (hunk 5 above)
Comment (TonyHan11) [reply core/arch/arm/plat-sam/conf.mk]: Yes, this pattern works. Updated and added the review tags. Thank you.

File: core/arch/arm/plat-sam/platform_sama7g5.c [hunk 6]
```diff
@@ -0,0 +1,932 @@
...
+	 */
+	srtop_setting =	MATRIX_SRTOP(4, MATRIX_SRTOP_VALUE_128M);
+	sasplit_setting = MATRIX_SASPLIT(4, MATRIX_SASPLIT_VALUE_128M);
+	ssr_setting = MATRIX_LANSECH_NS(4) |
+		      MATRIX_RDNSECH_NS(4) |
+		      MATRIX_WRNSECH_NS(4);
+	matrix_configure_slave_security(base, 7, srtop_setting,
+					sasplit_setting, ssr_setting);
+
+	/* 8: NFC_RAM
```
[the remark this answers is not in the document]
Comment (TonyHan11) [reply core/arch/arm/plat-sam/platform_sama7g5.c]: Updated and replaced `COMPILE_TIME_ASSERT()` with `static_assert()`, thanks.

File: core/arch/arm/plat-sam/platform_sama7g5.c [hunk 7]
```diff
@@ -0,0 +1,935 @@
...
+#include <assert.h>
+#include <console.h>
+#include <drivers/atmel_uart.h>
+#include <drivers/gic.h>
+#include <drivers/tzc400.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <matrix.h>
+#include <platform_config.h>
+#include <tz_matrix.h>
```
Comment (etienne-lms) [inline core/arch/arm/plat-sam/platform_sama7g5.c:15]: missing `include <mm/core_mmu.h>` and `#include <util.h>`.

File: core/arch/arm/plat-sam/platform_sama7g5.c [hunk 8]
```diff
@@ -0,0 +1,935 @@
...
+			CORE_MMU_PGDIR_SIZE);
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, MATRIX_BASE_ADDRESS,
+			CORE_MMU_PGDIR_SIZE);
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, TZC_BASE_ADDRESS, CORE_MMU_PGDIR_SIZE);
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, TZPM_BASE_ADDRESS,
+			CORE_MMU_PGDIR_SIZE);
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, GIC_INTERFACE_BASE, GICC_SIZE);
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, GIC_DISTRIBUTOR_BASE, GICD_SIZE);
+
+static struct matrix matriies[] = {
```
Comment (etienne-lms) [inline core/arch/arm/plat-sam/platform_sama7g5.c:29]: nitpicking: prefer `matrices` or `matrixes` over `matriies`

**Review by etienne-lms — COMMENTED** [review]
2 last comments.
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` once addressed.

File: core/arch/arm/plat-sam/platform_sama7g5.c (hunk 8 above)
Comment (TonyHan11) [reply core/arch/arm/plat-sam/platform_sama7g5.c]: Updated with review tags added, thanks.

**Pull request merged**